source/common/upstream/outlier_detection_impl.cc

```diff
@@ -208,12 +234,89 @@ void DetectorImpl::onConsecutive5xxWorker(HostSharedPtr host) {
   ejectHost(host, EjectionType::Consecutive5xx);
 }
 
+const double Utility::SUCCESS_RATE_STDEV_FACTOR = 1.9;
+
+double Utility::successRateEjectionThreshold(double success_rate_sum,
+                                             std::vector<double>& success_rate_data) {
+  // This function is using mean and standard deviation as statistical measures for outlier
+  // detection. First the mean is calculated by dividing the sum of success rate data over the
+  // number of data points. Then variance is calculated by taking the mean of the
+  // squared difference of data points to the mean of the data. Then standard deviation is
+  // calculated by taking the square root of the variance. Then the outlier threshold is
+  // calculated as the difference between the mean and the product of the standard
+  // deviation and a constant factor.
+  //
+  // For example with a data set that looks like success_rate_data = {50, 100, 100, 100, 100} the
+  // math would work as follows:
+  // success_rate_sum = 450
+  // mean = 90
+  // variance = 400
+  // stdev = 20
+  // threshold returned = 52
+  double mean = success_rate_sum / success_rate_data.size();
+  double variance = 0;
+  std::for_each(success_rate_data.begin(), success_rate_data.end(),
+                [&variance, mean](double& v) { variance += std::pow(v - mean, 2); });
+  variance /= success_rate_data.size();
+  double stdev = std::sqrt(variance);
+
+  return mean - (SUCCESS_RATE_STDEV_FACTOR * stdev);
+}
+
+void DetectorImpl::successRateEjections() {
+  std::unordered_map<HostSharedPtr, double> valid_success_rate_hosts;
```

Comment:
You don't actually need map lookup here, an std::list or reserved std::vector containing a struct with host and double will be more efficient.